migrations/versions/a528c3c81506_unique_generic_sensor_ids.py

```diff
@@ -0,0 +1,405 @@
+        "weather_sensor",
+        "sensor",
+        ["id"],
+        ["id"],
+    )
+
+
+def downgrade():
+    downgrade_schema_old_sensors()
+    downgrade_data()
+    downgrade_schema_new_sensors()
+
+
+def downgrade_schema_old_sensors():
+    """Schema migration to decouple the id of old sensor tables from the new sensor table."""
+    op.drop_constraint("asset_id_sensor_fkey", "asset", type_="foreignkey")
+    op.drop_constraint("market_id_sensor_fkey", "market", type_="foreignkey")
+    op.drop_constraint(
+        "weather_sensor_id_sensor_fkey", "weather_sensor", type_="foreignkey"
+    )
+
+
+def downgrade_data():
+    """Data migration to retrieve the ids of old sensors.
+
+    Note that downgraded ids are not guaranteed to be the same as during upgrade."""
+
+    # Upgrade schema to support data downgrade
+    temporary_schema_upgrades()
+
+    # Declare ORM table views
+    t_markets = sa.Table(
+        "market",
+        sa.MetaData(),
+        sa.Column("id", sa.Integer),
+        sa.Column("name", sa.String(80)),
+    )
+
+    # Use Alchemy's connection and transaction to go through the data
+    connection = op.get_bind()
+
+    # Get the max id used by assets and markets
+    max_asset_id = get_max_id(
+        connection, "asset"
+    )  # may be different than during upgrade!
+    max_market_id = get_max_id(
+        connection, "market"
+    )  # may be different than during upgrade!
+
+    # Select all existing ids that need migrating
+    market_results = connection.execute(
+        sa.select(
+            [
+                t_markets.c.id,
+                t_markets.c.name,
+            ]
+        )
+    ).fetchall()
+    # Iterate over all selected data tuples
+    for id_, name in market_results:
+        # Determine the new id
+        new_id = id_ - max_asset_id
+        # Update the id
+        connection.execute(
+            t_markets.update().where(t_markets.c.name == name).values(id=new_id)
+        )
+
+    # Repeat steps for weather sensors
+    t_weather_sensors = sa.Table(
+        "weather_sensor",
+        sa.MetaData(),
+        sa.Column("id", sa.Integer),
+        sa.Column("name", sa.String(80)),
+    )
+    weather_sensor_results = connection.execute(
+        sa.select(
+            [
+                t_weather_sensors.c.id,
+                t_weather_sensors.c.name,
+            ]
+        )
+    ).fetchall()
+    for id_, name in weather_sensor_results:
+        # Determine the new id
+        new_id = id_ - max_market_id
+        # Update the id
+        connection.execute(
+            t_weather_sensors.update()
+            .where(t_weather_sensors.c.name == name)
+            .values(id=new_id)
+        )
+
+    # Downgrade schema after supporting data downgrade
+    temporary_schema_downgrades()
+
+
+def downgrade_schema_new_sensors():
+    """Schema migration to drop the new sensor table."""
+    op.drop_table("sensor")
+
+
+def temporary_schema_upgrades():
```

Comment:
Actually, this function and the one after that could be the same function (e.g. "`recreate_sensor_fks`"), with a parameter `recreate_with_cascade_onupdate: bool`.